crates/ruff_linter/src/suppression.rs

```diff
@@ -0,0 +1,905 @@
+        self.eat_whitespace();
+        if !self.cursor.eat_char('[') {
+            return self.error(ParseErrorKind::MissingCodes);
+        }
+
+        let mut codes: SmallVec<[TextRange; 2]> = smallvec![];
+
+        loop {
+            if self.cursor.is_eof() {
+                return self.error(ParseErrorKind::MissingBracket);
+            }
+
+            self.eat_whitespace();
+
+            if self.cursor.eat_char(']') {
+                break Ok(codes);
+            }
+
+            let code_start = self.offset();
+            if !self.eat_word() {
+                return self.error(ParseErrorKind::InvalidCode);
+            }
+
+            codes.push(TextRange::new(code_start, self.offset()));
+
+            self.eat_whitespace();
+            if !self.cursor.eat_char(',') {
+                if self.cursor.eat_char(']') {
+                    break Ok(codes);
+                }
+
+                return if self.cursor.is_eof() {
+                    self.error(ParseErrorKind::MissingBracket)
+                } else {
+                    self.error(ParseErrorKind::MissingComma)
+                };
+            }
+        }
+    }
+
+    fn eat_whitespace(&mut self) -> bool {
+        if self.cursor.eat_if(char::is_whitespace) {
+            self.cursor.eat_while(char::is_whitespace);
+            true
+        } else {
+            false
+        }
+    }
+
+    fn eat_word(&mut self) -> bool {
+        if self.cursor.eat_if(char::is_alphabetic) {
+            // Allow `:` for better error recovery when someone uses `lint:code` instead of just `code`.
+            self.cursor
+                .eat_while(|c| c.is_alphanumeric() || matches!(c, '_' | '-' | ':'));
+            true
+        } else {
+            false
+        }
+    }
+
+    fn offset(&self) -> TextSize {
+        self.range.start() + self.range.len() - self.cursor.text_len()
+    }
+
+    fn error<T>(&self, kind: ParseErrorKind) -> Result<T, ParseError> {
+        Err(ParseError::new(
+            kind,
+            TextRange::new(self.offset(), self.range.end()),
+        ))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use insta::assert_debug_snapshot;
+    use ruff_python_parser::{Mode, ParseOptions, parse};
+    use ruff_text_size::{TextRange, TextSize};
+    use similar::DiffableStr;
+
+    use crate::suppression::{
+        ParseError, SuppressionAction, SuppressionComment, SuppressionParser, Suppressions,
+    };
+
+    #[test]
+    fn suppressions_from_tokens() {
+        let source = "
+# ruff: noqa  # ignored
+
+# comment here
+print('hello')  # ruff: disable[phi]  # trailing
+
+# ruff: disable[alpha]
+def foo():
+    # ruff: disable[beta,gamma]
+    if True:
+        # ruff: disable[delta]  # unmatched
+        pass
+    # ruff: enable[beta,gamma]
+# ruff: enable[alpha]
+
+# ruff: disable  # parse error!
+def bar():
+    # ruff: disable[zeta]  # unmatched
+    pass
+# ruff: enable[zeta]  # underindented
+    pass
+";
```

Comment:
I think it would be great to have a few more tests here:

* What if we have to succinct `ruff: enable`
* What about multiple `ruff: disable` and only one `ruff: enable`
* What if the `ruff: disable` and `ruff: enable` match but the codes are in a different order?